tianshou/policy/modelfree/discrete_sac.py

```diff
@@ -0,0 +1,144 @@
+    """Implementation of SAC for Discrete Action Settings. arXiv:1910.07207.
+
+    :param torch.nn.Module actor: the actor network following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer actor_optim: the optimizer for actor network.
+    :param torch.nn.Module critic1: the first critic network. (s -> Q(s))
+    :param torch.optim.Optimizer critic1_optim: the optimizer for the first
+        critic network.
+    :param torch.nn.Module critic2: the second critic network. (s -> Q(s))
+    :param torch.optim.Optimizer critic2_optim: the optimizer for the second
+        critic network.
+    :param float tau: param for soft update of the target network, defaults to
+        0.005.
+    :param float gamma: discount factor, in [0, 1], defaults to 0.99.
+    :param (float, torch.Tensor, torch.optim.Optimizer) or float alpha: entropy
+        regularization coefficient, default to 0.2.
+        If a tuple (target_entropy, log_alpha, alpha_optim) is provided, then
+        alpha is automatatically tuned.
+    :param bool reward_normalization: normalize the reward to Normal(0, 1),
+        defaults to ``False``.
+    :param bool ignore_done: ignore the done flag while training the policy,
+        defaults to ``False``.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        actor: torch.nn.Module,
+        actor_optim: torch.optim.Optimizer,
+        critic1: torch.nn.Module,
+        critic1_optim: torch.optim.Optimizer,
+        critic2: torch.nn.Module,
+        critic2_optim: torch.optim.Optimizer,
+        tau: float = 0.005,
+        gamma: float = 0.99,
+        alpha: Union[
+            float, Tuple[float, torch.Tensor, torch.optim.Optimizer]
+        ] = 0.2,
+        reward_normalization: bool = False,
+        ignore_done: bool = False,
+        estimation_step: int = 1,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(actor, actor_optim, critic1, critic1_optim, critic2,
+                         critic2_optim, [-np.inf, np.inf], tau, gamma, alpha,
+                         reward_normalization, ignore_done, estimation_step,
+                         **kwargs)
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        obs = getattr(batch, input)
+        logits, h = self.actor(obs, state=state, info=batch.info)
+        dist = Categorical(logits=logits)
+        act = dist.sample()
+        return Batch(logits=logits, act=act, state=h, dist=dist)
+
+    def _target_q(
+        self, buffer: ReplayBuffer, indice: np.ndarray
+    ) -> torch.Tensor:
+        batch = buffer[indice]  # batch.obs: s_{t+n}
+        with torch.no_grad():
+            obs_next_result = self(batch, input="obs_next")
```

Comment:
I would rather use `self.forward` since calling `self` directly in internal methods is usually.

Reply:
It's not recommended to use `self.forward` directly because the internal `__call__` function manages all registered hooks.  Calling `self.forward` directly would skip those hooks. [https://discuss.pytorch.org/t/any-different-between-model-input-and-model-forward-input/3690/2](url)

Reply:
I didn't know there was any difference between `__call__` and `forward`. Ok then !